### PR TITLE
Feature/gdal datasources

### PR DESCRIPTION
**Pull request by milos-colic**

### CODE REVIEW

**Comment by edurdevic** [comment]
@milos-colic can we please split this into two separate PRs?
I see that the grid_distance has a low coverage, and will break the custom grid PR. We can merge the raster part now and take the necessary time to add the distance for all grids after we have the custom one.

**Comment by milos-colic** [comment]
@edurdevic I wouldn't want to split this into 2, for 1 grid distance is properly covered but the pluging we use doesnt work well, I did a correct testing for it, 2nd without this distance function grid reader for rasters wont work outside of photon and that is a na issue. So I would rather accept this and more custom grid or add custom grid distance function (BNG has exact same distance fucntion as a custom grid, so it should be rather fast to add ir)

File: src/test/scala/com/databricks/labs/mosaic/expressions/index/CellKLoopBehaviors.scala [hunk 1]
```diff
@@ -70,6 +70,10 @@ trait CellKLoopBehaviors extends MosaicSpatialQueryTest {
         )
 
         an[Error] should be thrownBy badExpr.inputTypes
+        mc.getIndexSystem match {
+            case H3IndexSystem  => noException should be thrownBy mc.getIndexSystem.kLoop(-1L, -1)
+            case BNGIndexSystem => an[Exception] should be thrownBy mc.getIndexSystem.kLoop(-1L, -1)
```
[the remark this answers is not in the document]
Comment (milos-colic) [reply src/test/scala/com/databricks/labs/mosaic/expressions/index/CellKLoopBehaviors.scala:75]: H3 should throw an exception as well but it doesnt and that is a native issue. Something we cant control. But for exmaple h3_kring works even if you pass -1 as the cell ID which is clearly wrong. We could make fail only if we had another wrapper around H3.

File: src/main/scala/com/databricks/labs/mosaic/expressions/raster/base/RasterToGridExpression.scala [hunk 2]
```diff
@@ -54,7 +56,11 @@ abstract class RasterToGridExpression[T <: Expression: ClassTag, P](
         val gt = raster.getRaster.GetGeoTransform()
         val resolution = arg1.asInstanceOf[Int]
         val bandTransform = (band: MosaicRasterBand) => {
-            val results = band.transformValues[(Long, Double)](pixelTransformer(gt, resolution), (0L, -1.0))
+            val pixelArea = raster.xSize * raster.ySize
+            val (originCellId, _) = pixelTransformer(gt, resolution, asCentroid = false)(0, 0, 0)
+            val cellArea = indexSystem.indexToGeometry(originCellId, geometryAPI).getArea
+            val asCentroidFlag = pixelArea >= cellArea
```
[the remark this answers is not in the document]
Comment (milos-colic) [reply src/main/scala/com/databricks/labs/mosaic/expressions/raster/base/RasterToGridExpression.scala:62]: No it should not, if the area of the raster pixel is smaller than the H3 cell then there are multiple H3 cells in one raster pixle so it should go through smoothing logic. if the area of H3 is larger than the raster pixel then it should be the centroid.

[the remark this answers is not in the document]
Comment (edurdevic) [reply src/main/scala/com/databricks/labs/mosaic/expressions/raster/base/RasterToGridExpression.scala]: We discussed this offline, pushed changes to default always to centroid

**Review by edurdevic — APPROVED** [review]
LGTM!

**Pull request merged**